File: backend-js/src/shared/relay/globalIdCodec.ts
import { DomainError } from '../errors/DomainError';

export interface DecodedGlobalId {
  type: string;
  id: string;
}

export function encodeGlobalId(type: string, id: string): string {
  return Buffer.from(`${type}:${id}`, 'utf8').toString('base64url');
}

export function decodeGlobalId(globalId: string): DecodedGlobalId {
  try {
    const decoded = Buffer.from(globalId, 'base64url').toString('utf8');
    const separator = decoded.indexOf(':');
    if (separator <= 0 || separator === decoded.length - 1) {
      throw DomainError.validation('Invalid global id format');
    }

    return {
      type: decoded.slice(0, separator),
      id: decoded.slice(separator + 1),
    };
  } catch (error) {
    if (error instanceof DomainError) {
      throw error;
    }
    throw DomainError.validation('Invalid global id encoding');
  }
}

File: backend-java/src/main/java/com/woovi/crudbank/shared/relay/GlobalIdCodec.java
package com.woovi.crudbank.shared.relay;

import com.woovi.crudbank.shared.error.DomainException;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

@Component
public class GlobalIdCodec {

    public String encode(String type, String id) {
        String raw = type + ":" + id;
        return Base64.getUrlEncoder().encodeToString(raw.getBytes(StandardCharsets.UTF_8));
    }

    public DecodedGlobalId decode(String globalId) {
        try {
            String decoded = new String(Base64.getUrlDecoder().decode(globalId), StandardCharsets.UTF_8);
            int separator = decoded.indexOf(':');
            if (separator <= 0 || separator == decoded.length() - 1) {
                throw DomainException.validation("Invalid global id format");
            }
            return new DecodedGlobalId(
                decoded.substring(0, separator),
                decoded.substring(separator + 1)
            );
        } catch (IllegalArgumentException ex) {
            throw DomainException.validation("Invalid global id encoding");
        }
    }

    public record DecodedGlobalId(String type, String id) {
    }
}
